### ISSUE TITLE
MHO-C401 new revision - display frozen

### ISSUE THREAD

**Issue by evlo** [pvvx/ATC_MiThermometer#276]
Hi, on MHO-C401 display is frozen, even for the not Azarton verison

![image](https://user-images.githubusercontent.com/8813196/206474516-cf33ba04-0222-49af-befa-576d5c9009ff.png)
![20221208_153047](https://user-images.githubusercontent.com/8813196/206474575-4033acc9-2fa7-4657-9dca-439c72c36a0e.jpg)
![20221208_153149](https://user-images.githubusercontent.com/8813196/206474602-a959dbd5-6f9d-4ebb-a044-b01d0aea4f06.jpg)

all display segments on
![20221208_154824](https://user-images.githubusercontent.com/8813196/206477295-55c0de95-d8c2-49ca-a37e-bc9b654e4fbf.jpg)

https://pvvx.github.io/MHO_C401/

display seems same as in https://github.com/znanev/MHO-C401

related https://github.com/pvvx/ATC_MiThermometer/issues/114

https://pvvx.github.io/ATC_MiThermometer/TelinkMiFlasher.html says Hardware Version: LYWSD03MMC B1.6, Software Version: 3.9, Sensor: SHTC3 so now I'm a bit confussed, isn't LYWSD03MMC  the small lcd one?

**Comment by cgernert** [comment]
I also ordered four new Mijia MMC, but inside the box are the Azarton ones, with Azarton display. Box comes with Chinese text, owl and all  😁
Seems they switched some things up.
![20221208_195045](https://user-images.githubusercontent.com/17023559/206542145-36e6a6e6-d936-416b-8c5e-ddd104a569c7.jpg)

**Comment by evlo** [comment]
Altrough even with this https://github.com/pvvx/ATC_MiThermometer/blob/master/Original_OTA_Azarton_MHO_C401_v1.0.0_0016.bin firmware screen is stuck.

**Comment by pvvx** [comment]
> Seems they switched some things up.

![image](https://user-images.githubusercontent.com/12629515/206698877-1075ac44-9bb5-4029-bcee-8185d50a2c68.png)
![image](https://user-images.githubusercontent.com/12629515/206699032-b2501c16-3ba3-4bc6-a2b9-3660cef4ba91.png)

**Comment by cgernert** [comment]
Maybe there will be a future display version with the top icon, I don't know.

But I did not want to distract from evlo's issue.
Perhaps it is another controller for the display, but why the original firmware is working then? Very strange.
The screwdriver marks on the PCB can't be unseen. Luckily, there seems to be nothing under it. You really wanted that battery out of there, hm? 😂

**Comment by evlo** [comment]
https://github.com/pvvx/TlsrComSwireWriter/issues/6

I don't get it, I think one is just bricked

**Comment by pvvx** [comment]
> https://pvvx.github.io/ATC_MiThermometer/TelinkMiFlasher.html says Hardware Version: LYWSD03MMC B1.6, Software Version: 3.9, Sensor: SHTC3 so now I'm a bit confussed, isn't LYWSD03MMC the small lcd one?

Why did you write firmware from Xiaomi LYWSD03MMC into it?

**Comment by evlo** [comment]
I dunno, this is what the https://pvvx.github.io/ATC_MiThermometer/TelinkMiFlasher.html put on the device, i noticed it too late, but whenever i tried after that it showed correct firmware as the original, maybe because before i was flashing LYWSD03MMC or maybe some stupid oversight on my side like with the read instead of write script

**Comment by pvvx** [comment]
Firmware from Xiaomi LYWSD03MMC will work in MHO-C401. But will not display on the display. OTA will work and you can flash any other firmware. 
It is impossible to flash firmware from MHO-C401 or CGG1 to other options. Nothing will work.

**Comment by evlo** [comment]
I have tried building https://github.com/kloemi/ATC_MiThermometer and flashing it to the device, that also results in frozen display, so probably really even different than [Azarton](https://github.com/kloemi/ATC_MiThermometerAzarton)

**Comment by kloemi** [comment]
That's how it should work. Strange that display controller was updated again ...

**Comment by cgernert** [comment]
> > Seems they switched some things up.
> 
> ![image](https://user-images.githubusercontent.com/12629515/206698877-1075ac44-9bb5-4029-bcee-8185d50a2c68.png) ![image](https://user-images.githubusercontent.com/12629515/206699032-b2501c16-3ba3-4bc6-a2b9-3660cef4ba91.png)

Boom, there it is. I also built and flashed kloemi's firmware.
I don't know if this symbol is always present on the Azartons
![azarton-bt3](https://user-images.githubusercontent.com/17023559/207133331-82518caf-7e50-4210-9e01-42730c0bdf70.png)

**Comment by evlo** [comment]
So for you display updates with kloemi's fork?

**Comment by cgernert** [comment]
> So for you display updates with kloemi's fork?

Yes, it does. But the hardware of the devices' I got seems to be Azarton or related. (X-battery on photos)
https://github.com/pvvx/ATC_MiThermometer/issues/276#issuecomment-1343186448

**Comment by pvvx** [comment]
To support Azatron, you need a [read original full flash](https://github.com/pvvx/ATC_MiThermometer/issues/114#issuecomment-979458611) and a description of all E-Ink segments.
[Samples description of all E-Ink segments.](https://github.com/pvvx/ATC_MiThermometer/blob/master/src/epd_cgg1_2022.c#L16)
The command to output to the E-ink screen buffer is 
**60**b0b1b2...b13
where b0 is byte0, b1 is byte1, ...

**Comment by evlo** [comment]
> > So for you display updates with kloemi's fork?
> 
> Yes, it does. But the hardware of the devices' I got seems to be Azarton or related. (X-battery on photos) [#276 (comment)](https://github.com/pvvx/ATC_MiThermometer/issues/276#issuecomment-1343186448)

Oh right i finally get the X thing, I was super confused by it but now I see that mine has flash or something instead of X.

BTW it does not even activate in Mi Home with region CN - it connects, but does not "verify" - it would be fun if somebody was doing fake version of these.

If display does work with the https://github.com/pvvx/ATC_MiThermometer/blob/master/Original_OTA_Xiaomi_MHO_C401_v1.0.0_0010.bin firmware is it really necesary to do the dump? I'm very bad with soldering, that is why I'm asking.

**Comment by kloemi** [comment]
I'm wondering that it works although the FPC is the same as in the old MH0_C401. 
Regarding the bluetooth symbols:
Azarton has both as shown in the documents, but I decided to use only the first at connected state.

And as mentioned before the display buffer seems not to contain a bit for the battery sign and the degree sign. No idea how to set them.

**Comment by pvvx** [comment]
> This is with ble_hacker.bin still flashed on the device

Be careful with ble_hacker - it is designed to read and write Flash, as well as a one-time delete function with a old firmware startup restore.

**Comment by pvvx** [comment]
> And as mentioned before the display buffer seems not to contain a bit for the battery sign and the degree sign. No idea how to set them.

Is the buffer size correct?

> The command to output to the E-ink screen buffer is
60b0b1b2...b13
where b0 is byte0, b1 is byte1, ...
https://github.com/pvvx/ATC_MiThermometer/issues/276#issuecomment-1347620089
60FFFFFFFFFFFFFFFFFFFFFFFFFFFF All On
600000000000000000000000000000 All Off

**Comment by evlo** [comment]
> [I have such models](https://pvvx.github.io/MHO_C401/) and works with MHO_vXX.
I have tried flashing `MHO_C401_v38.bin` onto the one I extracted that FW from and it works flawlessly, i thought i received both same units, but clearly not, would be helpful to extract firmware from the one where the display freezes after flashing it with `MHO_C401_v38.bin`, but is working after flashing it with `Original_OTA_Xiaomi_MHO_C401_v1.0.0_0010.bin` using BLE_hacker or it would just basically get  `Original_OTA_Xiaomi_MHO_C401_v1.0.0_0010.bin` again?

**Comment by pvvx** [comment]
I don't really understand what you are describing. BLE_hacker has the option to restore the launch of the previous firmware. But this button should be selected only once and disconnect the connection - disconnect...

**Comment by kloemi** [comment]
> Is the buffer size correct?

Hmm you're right. Don't understand why I reduced it to 14 byte but in my recordings it seems to have 16 byte. Going to test.

**Comment by evlo** [comment]
Hmmm, i thought the files would be identical, but there are some differences between two FF exports
https://1drv.ms/u/s!Ard77GaIVWRslMh15PZUTwMyzcdNkQ?e=FHuISp

***
_I have bought two MHO-C401

one is the one giving me issues, that i can flash using wired connection (if it feels like getting flashed) - it is very particular and even likes one ch340 flasher better then other (maybe i could try lowering baud rate) - I have fixed all my soldering. But I have tried soldering the RST pin, impossible for me, mayb ei have messed up something in the process.

from the non problematic one i uploaded extracted firmware from it in the post https://github.com/pvvx/ATC_MiThermometer/issues/276#issuecomment-1348813227

problematic one works with `Original_OTA_Xiaomi_MHO_C401_v1.0.0_0010.bin` firmware, but i can't connect it to mi app even - again, I could have messed up some soldering etc.

With `Original_OTA_Xiaomi_MHO_C401_v1.0.0_0010.bin` firmware it connects with [[Telink Flasher v4.7](https://pvvx.github.io/ATC_MiThermometer/TelinkMiFlasher.html)](https://pvvx.github.io/ATC_MiThermometer/TelinkMiFlasher.html) , i have uploaded `BLE_hacker` to it, then connected using [[Telink OTA Hacker](https://pvvx.github.io/CGPR1/TLSR825xOTA_Hacker.html)](https://pvvx.github.io/CGPR1/TLSR825xOTA_Hacker.html) - that I think would only restore `Original_OTA_Xiaomi_MHO_C401_v1.0.0_0010.bin` I flashed to it that before, but the two files are not exactly the same - maybe just different tokens etc.

Honestly I now think that it is just MHO C401 with HW defect, no new version or anything, now I even have issues flashing it when I soldered wires nicely and using different usb uart adapters.

I'm curious if the extract is the same or not, especially since i flashed `Original_OTA_Xiaomi_MHO_C401_v1.0.0_0010.bin` onto it before. Maybe i do not understand how the flashing works in this case

Weirdly flashing work whether whichever way i connect rx and tx - rx to rx or rx to tx with 1k ohm resistor in the way_

**Comment by pvvx** [comment]
The firmware azarton_mho-c410_ff.bin has only 87332 bytes from the beginning of Flash. And in the posted azarton_mho-c410_ff.bin, the firmware is intact and not changed.

**Comment by pvvx** [comment]
I made a mistake: compared one, wrote another:
```
> fc /b "ff (problematic).bin" Original_OTA_Xiaomi_MHO_C401_v1.0.0_0010.bin 
Comparison files ff (problematic).bin and ORIGINAL_OTA_XIAOMI_MHO_C401_V1.0.0_0010.BIN
FC: ff (problematic).bin is longer than ORIGINAL_OTA_XIAOMI_MHO_C401_V1.0.0_0010.BIN
```

**Comment by kloemi** [comment]
Increased buffer size to 16 did not fix it :-(

**Comment by cgernert** [comment]
> Increased buffer size to 16 did not fix it :-(

Non-Azarton has 18, right? But Azarton has additional symbols compared to Non-Azarton.
Could you also please enable issues on your repo? Or should we create a new Issue on this repo, because this issue is originally based on evlo's Non-Azarzon freezing unit.

**Comment by evlo** [comment]
Please dom, or continue this one https://github.com/pvvx/ATC_MiThermometer/issues/114 ?

If i understand correctly that one MHO-C401 does indeed have different firmware even when flash ORIGINAL_OTA_XIAOMI_MHO_C401_V1.0.0_0010.BIN onto it, so at this point mixing Azarton issues with it is really confusing.

But for that device I have that is having these issues, it is most likely not really worth investigating, until more turn out ...

**Comment by MattXcz** [comment]
Today I flashed 3x MHO_C401 , all are still frozen =(

**Comment by MattXcz** [comment]
Hi, I received 3 from Aliexpress today - all with new display version.

**Comment by juskalalie** [comment]
> I also ordered four new Mijia Miaomiaoce MMC, but inside the box are the Azarton ones, with Azarton display. Box comes with Chinese text, owl and all 😁 Seems they switched some things up. ![20221208_195045](https://user-images.githubusercontent.com/17023559/206542145-36e6a6e6-d936-416b-8c5e-ddd104a569c7.jpg) ![20221208_2026282](https://user-images.githubusercontent.com/17023559/206556648-c55ce587-57d9-40a8-9e14-782178ab5769.jpg)

Got 4 copies exactly like yours after ordering from the Mijia SmartHome Store on AliExpress
https://www.aliexpress.com/store/2185008

**Comment by MattXcz** [comment]
Mine was from "M-i M-ijia 18 Store"

**Comment by pvvx** [comment]
> Got 4 copies exactly like yours after ordering from the Mijia SmartHome Store on AliExpress
> https://www.aliexpress.com/store/2185008

I ordered, waiting for delivery (weeks 3) ...

**Comment by juskalalie** [comment]
Just to check, I flashed them with custom firmware -> screen got stuck. Was able to unstuck screen when flashing back to the original Azarton firmware available here.

**Comment by pvvx** [comment]
![image](https://user-images.githubusercontent.com/12629515/220659166-3698db0b-f89f-452c-8cb5-1e60500c605b.png)

Several [MHO-C401s (2022.11)](https://pvvx.github.io/MHO_C401N/) have arrived.

**Comment by cgernert** [comment]
> 

nice!

**Comment by pvvx** [comment]
Added support ver MHO-C401 (2022.11). 
Released beta version 4.2. Tested - all ok.

**Comment by juskalalie** [comment]
Wow that was blazing fast!!
Many thanks for your efforts! 👍

**Comment by pvvx** [comment]
The opportunity to make it faster was created thanks to @kloemi. He found how the signals from the SoC are distributed to the E-Ink driver chip. I did not even have to open the device, and I put the code from the CGG1 driver. The only thing I had to do was create a [map of the E-Ink segments](https://github.com/pvvx/ATC_MiThermometer/blob/master/src/epd_mho_c401n.c#L17).

**Comment by MattXcz** [comment]
Hello @pvvx  - I updated it to 4.2 beta, but sadly it still not update numbers on the display

**Comment by MattXcz** [comment]
> > Hello @pvvx - I updated it to 4.2 beta, but sadly it still not update numbers on the display
> 
> Try the MHO_C401**N**_v42.bin

Yes. That did it. Now it works :) thank you !

**Comment by homepersonalprojects** [comment]
> [pvvx/TlsrComSwireWriter#6](https://github.com/pvvx/TlsrComSwireWriter/issues/6)
> 
> I don't get it, I think one is just bricked, others seem to have succeeded with that tool
> 
> I think the marks on pcb are from the factory sticker under battery
> 
> ![20221209_134433](https://user-images.githubusercontent.com/8813196/206718876-5925b44b-63c8-4edf-aa2c-f0ac3a592b69.jpg) ![20221209_134218](https://user-images.githubusercontent.com/8813196/206719106-5b115be2-4fd5-4e64-be1f-d21fa9d868f2.jpg)
> 
> I think there is no chance I can solder the reset with my skills anyway

How do you reset at the hardware level? It's a shame that both the units I have got bricked in 2 different ways. One doesn't simply turn on and another connects to ZMA but display is frozen.
